### ISSUE TITLE
How best to get timezone data for imported records

### ISSUE THREAD

**Issue by gene1wood** [owntracks/recorder#515]
I'm in the process of migrating from Google Timeline to Owntracks Recorder. It [sounds like](https://www.technowizardry.net/2024/01/migrating-from-google-location-history-to-owntracks/#importing-into-owntracks) using the importer isn't ideal and so I've been following that blog entry (and others who are doing the same) to extend and improve that python import code to consume the Google Timeline JSON and render it into Owntracks `rec` files.

This has worked well. And I want to acknowledge that this method of importing historical data is not supported (as it's not going through the recorder API and instead just being written to the store files directly)

My question is, what would the best way be to get timezones working on this historical data. I've noticed that data points that are in the same timezone that I'm in do render correctly. I'm assuming this is because

* I'm adding new records from the Android app through the recorder API
* The recorder is doing the correct reverse geocoding and capturing in the LMDB database timezone information on those new records
* The recorder API understands that these historical data points are "near" to the new data points and so they must have the same timezone

or maybe the recorder just shows times in the browser's local timezone when it doesn't know the timezone of the data point?

(correct me if I'm wrong)

The problem with missing this timezone data for historical records is that it looks like they just display in my local timezone which makes the time of day for the data points displayed incorrect.

I know I can't do reverse geocoding on 13 years of data. I'm wondering if there's a way to seed the minimal number of historical reverse geocoded data points in other timezones so that all of the other historical data points in those timezones show up with the correct timezone correction.

**Comment by jpmens** [comment]
Closing as responded to.

**Comment by gene1wood** [comment]
> There is no relationship between time zone and the LMDB database. The latter is used for reverse geocoded information

@jpmens, ok, got it, thank you. I had assumed that the method for finding what timezone to display a point in was from the reverse geocode.

So I think my question is, when viewing owntracks data through the web UI that's been generated by the owntracks client it shows the timestamp for a given point in this format

```
2/1/2024, 10:47:27 PM
2024-02-01T22:47:27-0800[America/Los_Angeles]
```

The timezone is correct.

When viewing a point that I've imported from Google Timeline using the method I describe above, the time field shows now timezone

```
 5/11/2022, 11:01:19 AM 
```

Where does the timezone value that's displayed come from so I can figure out what I'm doing wrong that's causing the historical imported data to not display a timezone?

The reason I'm asking is that for points which are in the timezone I live in, everything is fine because they display a time which is accurate even if the timezone isn't explicitly displayed. 

But for points that are in other timezones (for example when I've been in another country in the past), the times all display in my "home" timezone, not the timezone that the point is located in. As a result, it shows travel from a house to a grocery store at 4am.

**Comment by jpmens** [comment]
I had to go back and [look it up](https://github.com/owntracks/recorder/issues/444#issuecomment-1918762335) myself

> When a new publish arrives, we perform reverse geo lookup on the lat/lon if configured. When using OpenCage (and only when using OpenCage), we obtain the timezone name from the annotations in their data. This we store alongside the actuall address data in our lmdb database (note tzname):

and I'm sorry that "lied" to you earlier:

> There is no relationship between time zone and the LMDB database

The first quote above is accurate.

Does this address your question?

**Comment by jpmens** [comment]
I have a VERY quick&dirty program [addtzname](https://github.com/owntracks/recorder/tree/master/contrib/addtzname) in contrib which  you could possibly use after importing your data to enrich the lmdb store with a timezone name (`tzname`) calculated from _tzdatadb_ which is also in `contrib/`.

**Comment by gene1wood** [comment]
@jpmens Thanks for the pointer

I've built addtzname in the [docker container](https://github.com/owntracks/docker-recorder).

This required adding some supporting packages

```
docker exec -it --user=root otrecorder sh
apk add --no-cache make musl-dev gcc
exit
```

then copying the recorder git repo into a mount available in the container and running `make` in the [`addtzname`](https://github.com/owntracks/recorder/tree/master/contrib/addtzname) directory

When I run the command in the addtzname README

```
ocat --storage /store --dump | /config/git-repo/contrib/addtzname/addtzname | ocat --storage /store/new-store --load
```

I get the error

```
Can't open zdb: No such file or directory
```

Looking in the code it looks like it can't load the TZDATADB

https://github.com/owntracks/recorder/blob/c2f0c43025f18c9010502f3f953915cde60e8c89/contrib/addtzname/addtzname.c#L36-L40

I can see that the TZDATADB is in the git-repo (`/config/git-repo/contrib/tzdatadb/timezone16.bin`) but I don't see it elsewhere in the docker container

Could you give me a pointer on how to get `addtzname` to be able to find the TZDATADB?

**Comment by gene1wood** [comment]
Ok, I see what my problem is.

The `addtzname` tool only operates on what `ocat --storage /path/to/store --dump` outputs.

And that ocat command output the ghash DB which contains the locations for which a reverse geo lookup has occurred. And in my case, all of the records that have had a reverse lookup already have a `tzname` value. It's the records for which a reverse lookup *hasn't* happened that I'm looking to add the `tzname.

For a large (decade) of data imported from Google Timeline, I don't have reverse lookup records for the points in ghash (to do so would require running the historical data against OpenCage which would either take a very long time or be very expensive). Shoot.

@jpmens Would it be true then that there's no way to enrich the imported data with `tzname` without doing an OpenCage reverse lookup on each location?

**Comment by jpmens** [comment]
The geohash is calculated from lat/lon at the moment the record is stored in the cache database. `ocat` dumps that out, and `addtzname` guesses the timezone from that data.

There is no need to perform reverse geo-lookups on the data.

**Comment by gene1wood** [comment]
@jpmens Ok, interesting

You'd said [above](https://github.com/owntracks/recorder/issues/515#issuecomment-2519441480)

> As long as you write the .rec files correctly, there is no reason why you shouldn't be creating them programmatically, that is perfectly acceptable.

Would there be a way, either while creating the `.rec` files, or with already created `.rec` files, to also populate the cache database with the lat/lon data points?

Maybe some command that would take a `.rec` file, extract the records, and create cache file records?

Then if I had the lat/lon data points in the cache file, when I run an `ocat --dump` it would send *all* of the lat/lon data points to `addtzname` (instead of what I have currently in the cache file which is only the lat/lon data points for locations created by the recorder, not any from my import where I created the historical `.rec` files)

**Comment by jpmens** [comment]
That sounds about right. Note that the geohash "width" should match what you've currently configured for the Recorder (possibly 7)

**Comment by jpmens** [comment]
Such a command doesn't exist. I was under the assumption that you would hack up your own. Sorry if I misinterpreted that.

**Comment by gene1wood** [comment]
Got it, I understand.

I wrote up a tool in Python and documented it here (along with my code that I used to convert from Google Timeline to Owntracks files). It's somewhat quick and dirty since I won't ever need to do this again after Google Timeline is no longer exportable from Google Takeout in 3 days.

https://github.com/gene1wood/convert_google_timeline_to_owntracks

@jpmens Thanks again for your help with this.
